vignettes/fims-path-maturity.Rmd

```diff
@@ -0,0 +1,549 @@
+
+```{r}          
+information$populations[[1]] <- list(id = 1, pointer = population_modules[[1]])
+information$populations[[2]] <- list(id = 2, pointer = population_modules[[2]])
+information$populations[[3]] <- list(id = 3, pointer = population_modules[[3]])
+
+information$maturity[[1]] <- list(id = 1, pointer = maturity_modules[[1]])
+information$maturity[[2]] <- list(id = 2, pointer = maturity_modules[[2]])
+```
+
+In FIMS, these definitions happen by passing information from the rcpp interface into FIMS Information. 
+```{r, eval = FALSE}
+## d0 is the pointer to information
+## lm0 is the pointer to a specific maturity module (e.g. lm0 = maturity_modules[[1]])
+# d0->maturity_models[lm0->id] = lm0;
+```
+
+We can then loop through each population module defined in information and set the maturity we're getting from the user specified in the Rcpp interface back out to the Population class used to run all the calculations in the model. After each line of code in the following chunk, corresponding C++ code from FIMS is directly below and commented out.
+
+```{r}
+for(it in 1:length(population_modules)){
+  # for (population_iterator it = this->populations.begin();
+  #            it != this->populations.end(); ++it) {
+  
+  p <- information$populations[[it]][2]
+  #std::shared_ptr<fims::Population<T> > p = (*it).second;
+  
+  maturity_uint <- p$pointer$maturity_id
+  #uint32_t maturity_uint = static_cast<uint32_t>(p->maturity_id);
+  
+  newit <- which(sapply(information$maturity, function(x) x$id) == maturity_uint)
+  #maturity_models_iterator it = this->maturity_models.find(maturity_uint)
+  
+  p$pointer$maturity #MaturityBase
+  p$pointer$maturity <- information$maturity[[newit]][[2]]
+  p$pointer$maturity #LogisticMaturity1
+  #p->maturity = (*it).second;
+  
+  population_modules[[it]][2]$maturity <- p$pointer$maturity
+}
+
+sapply(population_modules, function(x) x[2]$maturity)
+```
+
+In R, the above code sets up an intermediate value, p. Once we update the maturity from p to equal the correct maturity module, `p$pointer$maturity <- information$maturity[[newit]][[2]]`, the above code passes this back to population_modules, `population_modules[[it]][2]$maturity <- p$pointer$maturity`. In this context, `population_modules` is the population class held in memory that is being used to run calculations.
+
+The deeper level of abstraction happening in C++ is that in FIMS, _p is the population module_ stored in memory. An equivalent line of code to `population_modules[[it]][2]$maturity <- p$pointer$maturity` isn't needed in FIMS because when we update p, we are also updating the population class held in memory that is being used to run calculations.
+
+## Clearing FIMS
+
+Because FIMS sets up objects stored in memory, simply running `rm(fims)`, won't free this memory back up. There is a function that can be run to clear out memory:
+```{r, eval = FALSE}
+fims$clear()
```

Comment:
when I run this line fims (and other things) still exists in my session. perhaps this is expected, but maybe some more description on why that is expected and ok would be good.

Reply:
I added a comment that rm(fims), rm(population) and rm(maturity) also need to be run in R to remove these objects from the R session

Reply:
maybe clear could be defined to do the rm commands on appropriate objects as well?

Reply:
clear is specifically a C++ function, so we wouldn't be able to call any R functions from inside it. We could create a wrapper than calls fims$clear and rm.

Reply:
How would we know what R objects to clear though because I don't think it is a requirement that users name things `fims`, `population`, and `maturity`. It would be much cleaner to develop things in an environment and then just remove the whole environment.